### PR TITLE
feat(ingest/nifi): ingest process group as browse path v2, incremental lineage

### PR DESCRIPTION
**Pull request by mayurinehate**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py [hunk 1]
```diff
@@ -123,7 +132,21 @@ class NifiSourceConfig(EnvConfigMixin):
...
     ca_file: Optional[Union[bool, str]] = Field(
         default=None,
-        description="Path to PEM file containing certs for the root CA(s) for the NiFi",
+        description="Path to PEM file containing certs for the root CA(s) for the NiFi."
+        "Set to False to disable SSL verification.",
+    )
+
+    # As of now, container entities retrieval does not respect browsePathsV2 similar to container aspect.
+    # Consider enabling this when entities with browsePathsV2 pointing to container also get listed in container entities.
+    emit_process_group_as_container: bool = Field(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/nifi.py:141]: this is only needed until we show browsePathV2 on entity cards right

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py [hunk 2]
```diff
@@ -1107,17 +1132,113 @@ def construct_job_workunits(
...
+            for inlet in inlets:
+                patch_builder.add_input_dataset(inlet)
+            for outlet in outlets:
+                patch_builder.add_output_dataset(outlet)
+            for inJob in inputJobs:
+                patch_builder.add_input_datajob(inJob)
+            for patch_mcp in patch_builder.build():
+                yield MetadataWorkUnit(
+                    id=f"{job_urn}-{patch_mcp.aspectName}", mcp_raw=patch_mcp
+                )
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/nifi.py:1154]: eventually I would like to move this into `auto_incremental_lineage`

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py [hunk 3]
```diff
@@ -1107,17 +1132,113 @@ def construct_job_workunits(
...
+            entityUrn=entity_urn,
+            aspect=BrowsePathsV2Class(
+                path=[
+                    BrowsePathEntryClass(id=flow_urn, urn=flow_urn),
+                    *self.get_browse_path_entries(process_group_id),
+                ]
             ),
         ).as_workunit()
 
+    def get_browse_path_entries(
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/nifi.py:1179]:
```suggestion
    def _get_browse_path_v2_entries(
```

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py [hunk 4]
```diff
@@ -1107,17 +1132,113 @@ def construct_job_workunits(
...
+            yield from self.construct_process_group_workunits(pg.parent_group_id)
+
+            if self._is_root_process_group(pg.parent_group_id):
+                yield self.gen_browse_path_v2_workunit(
+                    container_key.as_urn(), pg.parent_group_id
+                )
+
+    def gen_process_group_key(self, process_group_id: str) -> ProcessGroupKey:
+        return ProcessGroupKey(
+            process_group_id=process_group_id, platform=NIFI, env=self.config.env
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/nifi.py:1239]: are `process_group_id`s globally unique within nifi? it seems like they're scoped to the flow

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py [hunk 5]
```diff
@@ -1107,17 +1132,113 @@ def construct_job_workunits(
...
+    def get_browse_path_entries(
+        self, process_group_id: str
+    ) -> List[BrowsePathEntryClass]:
+        """Browse path entries till current process group"""
+        if self._is_root_process_group(process_group_id):
+            return []
+
+        current_process_group = self.nifi_flow.processGroups[process_group_id]
+        parent_pg_id = current_process_group.parent_group_id
+        if parent_pg_id and not self._is_root_process_group(parent_pg_id):
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/nifi.py:1188]: the `_is_root_process_group` check is already done in the recursive call - imo it's clearer to not check it again here

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py [hunk 6]
```diff
@@ -392,6 +400,10 @@ def __init__(self, config: NifiSourceConfig, ctx: PipelineContext) -> None:
         if self.config.ca_file is not None:
             self.session.verify = self.config.ca_file
 
+        # To keep track of process groups (containers) which have already been ingested
+        # Required, as we do not ingest all process groups but only those that have known ingress/egress processors
+        self.processed_pgs: List[str] = []
```
Comment (hsheth2) [inline metadata-ingestion/src/datahub/ingestion/source/nifi.py:405]: this is only used when `emit_process_group_as_container` is enabled right?

**Review by hsheth2 — APPROVED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py (hunk 1 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/nifi.py:141]: yes

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py (hunk 2 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/nifi.py:1154]: right!

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py (hunk 4 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/nifi.py:1239]: I think, this is safe.
For a single nifi clustered or standalone deployment - process group id would be unique. As of now - entire deployment is represented as single dataflow - for nothing else comes closer.
In future when multiple nifi deployments come in picture (seems rare, given the complexity of single nifi deployment), addition of platform-instance capability in nifi source should handle the uniqueness.

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py (hunk 6 above)
Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/nifi.py:405]: yes

File: metadata-ingestion/src/datahub/ingestion/source/nifi.py (hunk 4 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/source/nifi.py:1240]: sounds good

**Comment by hsheth2** [comment]
CI failures are related to https://github.com/datahub-project/datahub/pull/10235

**Pull request merged**